test/github_pullrequest_retest_test.go

```diff
@@ -51,38 +52,84 @@ func TestGithubPullRequestRetest(t *testing.T) {
 	assert.Assert(t, repo.Status[len(repo.Status)-1].Conditions[0].Status == corev1.ConditionTrue)
 }
 
-func TestGithubPullRequestRetestForParticularPipelineRun(t *testing.T) {
-	ctx := context.TODO()
+func TestGithubPullRequestGitOpsComments(t *testing.T) {
+	ctx := context.Background()
 	runcnx, ghcnx, opts, targetNS, targetRefName, prNumber, sha := tgithub.RunPullRequest(ctx, t,
 		"Github Retest comment", []string{"testdata/pipelinerun.yaml", "testdata/pipelinerun-gitops.yaml"}, false)
 	defer tgithub.TearDown(ctx, t, runcnx, ghcnx, prNumber, targetRefName, targetNS, opts)
 
-	runcnx.Clients.Log.Infof("Creating /retest pipelinerun-name in PullRequest")
-	_, _, err := ghcnx.Client.Issues.CreateComment(ctx,
-		opts.Organization,
-		opts.Repo, prNumber,
-		&github.IssueComment{Body: github.String("/retest pr-gitops-comment")})
+	pruns, err := runcnx.Clients.Tekton.TektonV1().PipelineRuns(targetNS).List(ctx, metav1.ListOptions{})
 	assert.NilError(t, err)
+	assert.Assert(t, len(pruns.Items) == 2)
 
-	runcnx.Clients.Log.Infof("Wait for the repository update to be updated")
-	waitOpts := twait.Opts{
-		RepoName:        targetNS,
-		Namespace:       targetNS,
-		MinNumberStatus: 3,
-		PollTimeout:     twait.DefaultTimeout,
-		TargetSHA:       sha,
+	tests := []struct {
+		name, comment string
+		prNum         int
+	}{
+		{
+			name:    "Retest",
+			comment: "/retest pr-gitops-comment",
+			prNum:   3,
+		},
+		{
+			name:    "Test and Cancel PipelineRun",
+			comment: "/cancel pr-gitops-comment",
+			prNum:   4,
+		},
 	}
-	err = twait.UntilRepositoryUpdated(ctx, runcnx.Clients, waitOpts)
-	assert.NilError(t, err)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			waitOpts := twait.Opts{
+				RepoName:        targetNS,
+				Namespace:       targetNS,
+				MinNumberStatus: tt.prNum,
+				PollTimeout:     twait.DefaultTimeout,
+				TargetSHA:       sha,
+			}
+			if tt.comment == "/cancel pr-gitops-comment" {
+				runcnx.Clients.Log.Info("/test pr-gitops-comment on Pull Request before canceling")
+				_, _, err := ghcnx.Client.Issues.CreateComment(ctx,
+					opts.Organization,
+					opts.Repo, prNumber,
+					&github.IssueComment{Body: github.String("/test pr-gitops-comment")})
+				assert.NilError(t, err)
 
-	runcnx.Clients.Log.Infof("Check if we have the repository set as succeeded")
-	repo, err := runcnx.Clients.PipelineAsCode.PipelinesascodeV1alpha1().Repositories(targetNS).Get(ctx, targetNS, metav1.GetOptions{})
-	assert.NilError(t, err)
-	assert.Assert(t, repo.Status[len(repo.Status)-1].Conditions[0].Status == corev1.ConditionTrue)
+				// max we will wait for 5 min
+				for time.Now().Before(time.Now().Add(5 * time.Minute)) {
```

Comment:
I didn't even know you can do in golang 🤔 

how does that work?

Reply:
if you can confirm that works properly then i'll merge it

but i wonder if it's just not wait for 5mn on every steps forever instead of waiting a max of 5mn

Reply:
> if you can confirm that works properly then i'll merge it

Yes it works

but Let me try with negative scenario one more time to be more sure

> but i wonder if it's just not wait for 5mn on every steps forever instead of waiting a max of 5mn

The loop will continue running until the current time is before 5 minutes from the initial time when the loop started. It doesn't wait for 5 minutes every time; rather, it checks the condition continuously until the specified time passed.

Reply:
I tried negative scenario and it did not work as expected

Updated code to use context with timeout